src/android/SecureStorage.java

```diff
@@ -66,192 +142,334 @@ public boolean execute(String action, CordovaArgs args, final CallbackContext ca
+
+        } else if (!RSA.isEntryAvailable(alias)) {
+            // Key for alias does not exist
+            handleLockScreen(IntentRequestType.INIT, service, callbackContext);
+
+        } else {
+            // No actions are required to init correctly
+            initSuccess(callbackContext);
         }
+        return true;
+    }
 
-        // Encrypt a message using the key with an alias associated with the store name
-        if ("encrypt_rsa".equals(action)) {
-            Log.v(TAG, "Called encrypt_rsa action");
-            final String service = args.getString(0);
-            final String encryptMe = args.getString(1);
-            cordova.getThreadPool().execute(new Runnable() {
-                public void run() {
-                    try {
-                        byte[] encrypted = RSA.encrypt(encryptMe.getBytes(), service2alias(service));
-                        callbackContext.success(Base64.encodeToString(encrypted, Base64.DEFAULT));
-                    } catch (Exception e) {
-                        Log.e(TAG, "Encrypt (RSA) failed :", e);
-                        callbackContext.error(e.getMessage());
-                    }
-                }
-            });
-            return true;
+    private boolean checkForSecurityMigration() {
+        Context ctx = getContext();
+        SharedPreferences preferences = ctx.getSharedPreferences(ctx.getPackageName() + "_SM",0);
+        String isMigrated = preferences.getString(MIGRATED_FOR_SECURITY, "");
+        //check OS then check if keys exist and migration was done
+        if(Build.VERSION.SDK_INT > Build.VERSION_CODES.M){
+
+            int size = initializePreferences();
+            Log.e(TAG, "TARGET OS");
+            if(isMigrated.equals("TRUE")){
+
+                Log.e(TAG, "MIGRATION ALREADY DONE");
+                return false;
+            }
+            //the target case of migration
+            else if(size > 0){
+
+                Log.e(TAG, "TARGET CASE FOR MIGRATION");
+                return true;
+            }
+            //new use, meaning we should put the tag in as to not trigger a unwanted migration
+            else{
+
+                Log.e(TAG, "NEW USE");
+                markAsMigrated(preferences);
+                return false;
+            }
         }
+        else{
+            //nothing to do return false
+            return false;
+        }
+    }
+
+    private void markAsMigrated(SharedPreferences preferences) {
+
+        SharedPreferences.Editor editor = preferences.edit();
+        editor.putString(MIGRATED_FOR_SECURITY, "TRUE");
+        editor.commit();
+    }
+
+    private int initializePreferences() {
+
+        Context ctx = getContext();
+        File prefdir = new File(ctx.getApplicationInfo().dataDir,"shared_prefs");
+        String[] filenames = prefdir.list();
+        int i = 0;
+        for(String name : filenames){
+            if(name.contains("SS")){
+                Log.e("BIGMIGRATION","FILENAMES: " + name);
```

Comment:
Once again, please do not use error level logs for regular operations. Search for usages of Log.e and change them (when they're not actually errors). Also use TAG as the tag.

I'd also say you should remove this entry, so we don't display paths to files with potentially sensitive data.

Reply:
The TAG value should be `TAG` instead, as that will allow developers consuming logs to understand where that log comes from.